doc/code/curves.md

```diff
@@ -0,0 +1,217 @@
+is the case for most RTS games - these downsides are usually outweight by the positives.
+
+## Architecture
+
+![Curve classes hierarchy UML](images/curve_classes_uml.svg)
+
+openage provides several curve types with different interpolation behaviour that each have
+their specific use case. Primitive data types are covered by the `BaseCurve` interface
+and its derivates of which the `Discrete`, `Continuous`, and `Segmented` are the practically
+usable types. For containers, there are classes for `Queue` and `UnorderedMap` available.
+All curve types are templated, so that they can store any type of value (with some constraints).
+The usable types are explained in more detail in the [Curve Types](#curve-types) section.
+
+Keyframes are implemented as time-value pairs by the `Keyframe` class. `KeyframeContainer`
+is used by all curves to manage their keyframe storage. It also provides functionality to
+efficiently access and insert keyframes as wel as sorting them by time. For the time
+type, the [simulation time type](/doc/code/time.md) from the `openage::time` namespace is
+used.
+
+It should be noted that curves are not useful in every situation as keyframe insertion,
+interpolation, and searching keyframes based on time create significant overhead. Curves
+should be used for variables or members where
+
+* data values must be tracked over time (e.g. HP of a game entity), or
+* data values are not modified/read very often (e.g. not every frame), or
+* data values are supposed to be sent over the network.
+
+This is usually the case for all game data in [game entity components](/doc/code/game_simulation/components.md)
+inside the game simulation.
+
+Curves should generally not be used for variables or members where
+
+* data values are not tracked over time (e.g. for temporary variables), or
+* data values are modified/read very often (e.g. in the rendering loop), or
+* data values don't affect the simulation state (e.g. visual settings).
+
+
+## Curve Types
+
+This section provides an overview over the available curves types.
+
+### Primitive
+
+![Primitive curve types UML](images/primitive_curves_uml.svg)
+
+Primitive curves are intended for members with single value types. These include, for example,
+the primitive C++ types (e.g. `int`, `float`, `std::string`), simple structs and data classes,
+and shared pointers.
+
+On contruction of a primitive curve object, a keyframe with time `t = time::time_t::min_value()`
+and the default value of the value type is inserted. This is done to ensure that for any
+requested time `t`, there is always a valid value to be returned. This mirrors the expected
+behaviour from declaring primitive values in C++ where members may be auto-initialized without
+explicit assignment to a default value. The default value for curves can also be explicitely
+assigned in the constructor. For value types that don't have default values, a default value
```

Comment:
types that don't have a default constructor need a default value passed

Reply:
fixed